execution_engine/src/core/engine_state/mod.rs

```diff
@@ -604,8 +617,96 @@ where
 
         let gas_limit = Gas::new(U512::from(std::u64::MAX));
 
+        let wasmless_transfer_gas_cost = Gas::new(U512::from(
+            protocol_data.system_config().wasmless_transfer_cost(),
+        ));
+
+        let wasmless_transfer_cost = match Motes::from_gas(
+            wasmless_transfer_gas_cost,
+            WASMLESS_TRANSFER_FIXED_GAS_PRICE,
+        ) {
+            Some(motes) => motes,
+            None => {
+                return Ok(ExecutionResult::precondition_failure(
+                    Error::GasConversionOverflow,
+                ))
+            }
+        };
+
+        let proposer_main_purse_balance_key = {
+            let proposer_main_purse = proposer_account.main_purse();
+
+            match tracking_copy
+                .borrow_mut()
+                .get_purse_balance_key(correlation_id, proposer_main_purse.into())
+            {
+                Ok(balance_key) => balance_key,
+                Err(_) => {
+                    return Ok(ExecutionResult::precondition_failure(
+                        Error::InsufficientPayment,
```

Comment:
This isn't actually `InsufficientPayment`, is it? Why not return the actual error instead of swallowing it?